In [3]:
import pandas as pd

# adjusting the path to the file is necessary, because the script is not executed in the same directory as the file
# import os
# print(os.getcwd()) # get current working directory to see where the script is executed

# Load the data into a DataFrame
file_path = 'Desktop/1.Semester/MachineLearing/Project/CO2/data/ESGEXCEL.csv' # please adjust the path to the file

try:
    data = pd.read_csv(file_path, delimiter=';', on_bad_lines='warn')
except pd.errors.ParserError as e:
    print(f"Error parsing CSV: {e}")

data.shape


/Users/frederikbomheuer


(16969, 68)

In [13]:
# Inspect the columns of the DataFrame
print("Columns in the DataFrame:", data.columns)
print("Column data types:")
data.dtypes

Columns in the DataFrame: Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022',
       '2023'],
      dtype='object')
Column data types:


Country Name      object
Country Code      object
Indicator Name    object
Indicator Code    object
1960              object
                   ...  
2019              object
2020              object
2021              object
2022              object
2023              object
Length: 68, dtype: object

In [12]:
# Display the first few rows of the transformed DataFrame
print("Overview of the first 5 rows in the DataFrame:")
data.head()

Overview of the first 5 rows in the DataFrame:


Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name     Indicator Code 1960  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS  NaN   
1            Access to electricity (% of population)     EG.ELC.ACCS.ZS  NaN   
2  Adjusted savings: natural resources depletion ...  NY.ADJ.DRES.GN.ZS  NaN   
3  Adjusted savings: net forest depletion (% of GNI)  NY.ADJ.DFOR.GN.ZS  NaN   
4                 Agricultural land (% of land area)     AG.LND.AGRI.ZS  NaN   

         1961         1962         1963         1964         1965  ...  \
0         NaN          NaN          NaN          NaN          NaN  ...   
1         NaN          NaN          NaN          NaN          NaN  ...   
2         NaN          NaN          NaN          NaN          NaN  ...   
3         NaN          NaN          NaN          NaN          NaN  ...   
4  30,9814141  30,98266305  31,00705428  31,01800095  31,04246564  ...   

          2014         2015         2016         2017         2018  \
0  86,70571737  86,94277798  87,22870518  87,39085552  87,61786211   
1  88,83227593  89,05385159  89,53901606  90,66275374  89,17693883   
2  10,05055446  6,130654513  5,265859073  6,245422284  8,187713883   
3  0,084360637  0,096672323  0,092911395  0,102683973  0,057123056   
4   39,8344215  39,87257452  39,93781393  39,98445203  39,96973753   

          2019         2020         2021 2022 2023  
0  87,79873981  87,94826443  88,09253583  NaN  NaN  
1  90,35280229  90,63505001  90,84566129  NaN  NaN  
2  7,234435527  4,598505988          NaN  NaN  NaN  
3  0,064515779  0,075685583          NaN  NaN  NaN  
4  39,90703091  39,97329048  39,97074236  NaN  NaN  

[5 rows x 68 columns]

In [4]:
print("Descriptive statistics of the columns:")
data.describe()

Descriptive statistics of the columns:


Country Name Country Code  \
count         16969        16969   
unique          239          239   
top      Arab World          ARB   
freq             71           71   

                                           Indicator Name  Indicator Code  \
count                                               16969           16969   
unique                                                 71              71   
top     Access to clean fuels and technologies for coo...  EG.CFT.ACCS.ZS   
freq                                                  239             239   

        1960  1961  1962  1963  1964  1965  ...   2014   2015   2016   2017  \
count   1623  2156  2165  2170  2175  2212  ...  12509  12390  11783  11577   
unique  1541  2078  2090  2089  2095  2126  ...  10724  10621  10136  10025   
top        0     0     0     0     0     0  ...      0      0      0      0   
freq      27    17    17    16    15    21  ...    372    367    266    271   

         2018   2019   2020  2021  2022         2023  
count   11710  11440  10731  7215  3640          453  
unique   9972   9988   9467  6334  3543          445  
top         0      0      0   100   100  6,235160271  
freq      280    265    268   178    15            3  

[4 rows x 68 columns]

In [28]:
# get more information about the columns
data['Indicator Name'].unique() 

array(['Access to clean fuels and technologies for cooking (% of population)',
       'Access to electricity (% of population)',
       'Adjusted savings: natural resources depletion (% of GNI)',
       'Adjusted savings: net forest depletion (% of GNI)',
       'Agricultural land (% of land area)',
       'Agriculture, forestry, and fishing, value added (% of GDP)',
       'Annual freshwater withdrawals, total (% of internal resources)',
       'Annualized average growth rate in per capita real survey mean consumption or income, total population (%)',
       'Cause of death, by communicable diseases and maternal, prenatal and nutrition conditions (% of total)',
       'Children in employment, total (% of children ages 7-14)',
       'CO2 emissions (metric tons per capita)', 'Coastal protection',
       'Control of Corruption: Estimate', 'Cooling Degree Days',
       'Economic and Social Rights Performance Score',
       'Electricity production from coal sources (% of total)',
       '

In [6]:
data['Indicator Code'].unique()

array(['EG.CFT.ACCS.ZS', 'EG.ELC.ACCS.ZS', 'NY.ADJ.DRES.GN.ZS',
       'NY.ADJ.DFOR.GN.ZS', 'AG.LND.AGRI.ZS', 'NV.AGR.TOTL.ZS',
       'ER.H2O.FWTL.ZS', 'SI.SPR.PCAP.ZG', 'SH.DTH.COMM.ZS',
       'SL.TLF.0714.ZS', 'EN.ATM.CO2E.PC', 'EN.CLC.CSTP.ZS', 'CC.EST',
       'EN.CLC.CDDY.XD', 'SD.ESR.PERF.XQ', 'EG.ELC.COAL.ZS',
       'EG.IMP.CONS.ZS', 'EG.EGY.PRIM.PP.KD', 'EG.USE.PCAP.KG.OE',
       'SP.DYN.TFRT.IN', 'AG.PRD.FOOD.XD', 'AG.LND.FRST.ZS',
       'EG.USE.COMM.FO.ZS', 'NY.GDP.MKTP.KD.ZG', 'EN.CLC.GHGR.MT.CE',
       'SI.POV.GINI', 'GE.EST', 'SE.XPD.TOTL.GB.ZS', 'EN.CLC.HEAT.XD',
       'EN.CLC.HDDY.XD', 'SH.MED.BEDS.ZS', 'SI.DST.FRST.20',
       'IT.NET.USER.ZS', 'SL.TLF.ACTI.ZS', 'EN.LND.LTMP.DC',
       'ER.H2O.FWST.ZS', 'SP.DYN.LE00.IN', 'SE.ADT.LITR.ZS',
       'EN.MAM.THRD.NO', 'EN.ATM.METH.PC', 'SH.DYN.MORT', 'SM.POP.NETM',
       'EN.ATM.NOXE.PC', 'IP.PAT.RESD', 'SH.H2O.SMDW.ZS',
       'SH.STA.SMSS.ZS', 'EN.ATM.PM25.MC.M3', 'PV.EST',
       'SP.POP.65UP.TO.ZS', 'EN.POP.DNST

In [7]:
data['Country Name'].unique()

array(['Arab World', 'Caribbean small states',
       'Central Europe and the Baltics', 'Early-demographic dividend',
       'East Asia & Pacific',
       'East Asia & Pacific (excluding high income)',
       'East Asia & Pacific (IDA & IBRD)', 'Euro area',
       'Europe & Central Asia',
       'Europe & Central Asia (excluding high income)',
       'Europe & Central Asia (IDA & IBRD)', 'European Union',
       'Fragile and conflict affected situations',
       'Heavily indebted poor countries (HIPC)', 'High income',
       'IBRD only', 'IDA & IBRD total', 'IDA blend', 'IDA only',
       'IDA total', 'Late-demographic dividend',
       'Latin America & Caribbean',
       'Latin America & Caribbean (excluding high income)',
       'Latin America & Caribbean (IDA & IBRD)',
       'Least developed countries: UN classification',
       'Low & middle income', 'Low income', 'Lower middle income',
       'Middle East & North Africa',
       'Middle East & North Africa (excluding high income)

Column data types:


Country Name      object
Country Code      object
Indicator Name    object
Indicator Code    object
1960              object
                   ...  
2019              object
2020              object
2021              object
2022              object
2023              object
Length: 68, dtype: object

**Findings from the global overview**

This global overview gives away the following facts about the available data:

- shape: 68 columns, 16969 rows
- all columns are of type "object" - neither numeric, nor string/text values
A certain amount of missing values, denoted both as NaN (not a number values)
- The columns represent key values such as country, but also the corresponding years and the indicator code/name
- The columns 'Country Code' and 'Indicator Code' do not give any information and are therefore obsolete
- The column 'Indicator Name' contains the country-specific features required for the analysis
- The names of the features in the column 'Indicator Name' are clear but too long

**Data Cleaning**

1. Removing the unnecessary columns "Country Code", "Indicator code"

In [30]:
# assign the data to a new DataFrame, which will be modified
data_clean = data

print("Original number of columns:")
print(data_clean.shape[1])

data_clean = data_clean.drop(['Country Code', 'Indicator Code'], axis='columns')

print("Current number of columns:")
print(data_clean.shape[1])

Original number of columns:
68
Current number of columns:
66


In [36]:
# Convert only the relevant columns to numeric types
numeric_columns = data_clean.columns[2:]  # the first four columns are non-numeric
data_clean[numeric_columns] = data_clean[numeric_columns].apply(pd.to_numeric, errors='coerce')

# Print the column data types after transformation
print("Column data types after transformation:")
print(data_clean.dtypes)

Column data types after transformation:
Country Name       object
Indicator Name     object
1960              float64
1961              float64
1962              float64
1963              float64
1964              float64
1965              float64
1966              float64
1967              float64
1968              float64
1969              float64
1970              float64
1971              float64
1972              float64
1973              float64
1974              float64
1975              float64
1976              float64
1977              float64
1978              float64
1979              float64
1980              float64
1981              float64
1982              float64
1983              float64
1984              float64
1985              float64
1986              float64
1987              float64
1988              float64
1989              float64
1990              float64
1991              float64
1992              float64
1993              float64
1994              float6

In [38]:
# save the feature names into a list of strings
chosen_cols = data_clean['Indicator Name'].unique()

# define an empty list, where sub-dataframes for each feature will be saved
frame_list = []

# iterate over all chosen features
for variable in chosen_cols:
    
    # pick only rows corresponding to the current feature
    frame = data_clean[data_clean['Indicator Name'] == variable]
    
    # melt all the values for all years into one column and rename the columns correspondingly
    frame = frame.melt(id_vars=['Country Name', 'Indicator Name']).rename(columns={'Country Code': 'country', 'variable': 'year', 'value': variable}).drop(['Indicator Name'], axis='columns')
    
    # add the melted dataframe for the current feature into the list
    frame_list.append(frame)


# merge all sub-frames into a single dataframe, making an outer binding on the key columns 'country','year'
from functools import reduce
all_vars = reduce(lambda left, right: pd.merge(left, right, on=['country','year'], how='outer'), frame_list)

KeyError: 'country'